### PR TITLE
Adds support for multidex APKs, plus some optimizations

### LINKED ISSUE #7: multidex support
Hello!

Looks like right now only the "classes.dex" file is being parsed :( Are there any plans to implement support for APKs that have multiple dex files inside?

Kindest regards,
Gabriel

### PR DESCRIPTION
**Pull request by iantruslove**
I had a crack at fixing the multidex issue. Fixes #7. 

Apologies for the meandering PR, I did some refactoring to learn about the code before digging in to the multidex issue itself.

- ae11dd7 through 862db8a, plus the other commits not explicitly mentioned below are some refactorings and minor fixes. The UTC time stamps is probably the most notable difference.
- 6a5d5c9 adds a bunch of pieces to build NinjaDroid into a docker image. I found this really helpful for testing - it's a stable environment for all the other pieces of APK tooling.
- 3ffaf4c fixes multidex. The changes aren't too complex, but I did make a minor adjustment to the Dex interface.
- 387bccf is a minor change that makes the run time about a minute faster for me.

Happy to make any changes to this, whatever you see fit.

### CODE REVIEW

**Comment by rovellipaolo** [comment]
Thanks a lot for this pull request! :)
I really appreciate it and encourage you to continue working on NinjaDroid in future.

Thus, I truly hope that you don't take it badly but, _next time_, please split each "new feature" or "standalone improvement" in a separate pull request. For example: one for the performance improvements, one for adding docker support, one for adding multidex support, ...
It's easier and safer to review X small pull requests rather than a single one X times bigger.

Anyway, I will review all the changes and let you know. ;)

**Comment by iantruslove** [comment]
> Thus, I truly hope that you don't take it badly but, _next time_, please split each "new feature" or "standalone improvement" in a separate pull request. 

Yeah, I know better than that... :) If you want I can split it up, it's probably easier for me to do that and you review the three or so MRs than the big monolith

**Comment by rovellipaolo** [comment]
Nah.. don't worry. I am already on it. :)

**Comment by iantruslove** [comment]
:D nice! 

I was just trying it out again on some more APKs but I think there's a problem retrieving resources from the zip archive. Digging in...

File: tests/test_aapt.py [hunk 1]
```diff
@@ -42,27 +42,22 @@ class TestAapt(unittest.TestCase):
...
+            if filename in self.files_properties:
+                self.files[filename] = join("tests", "data", filename)
 
     def tearDown(self):
         pass
 
     def test_get_app_name(self):
         for filename in self.files:
-            self.assertTrue(Aapt.get_app_name(self.files[filename]) == self.files_properties[filename]["app_name"])
+            self.assertEqual(
```
Comment (rovellipaolo) [inline tests/test_aapt.py]: Ops.. thanks for this change! :)

**Comment by iantruslove** [comment]
@rovellipaolo thanks for merging! Couple of things:

1) why did you revert the memoization commit? It significantly sped up processing. 
2) there were a couple of changes I made that you put back. There were valid reasons for those changes, albeit unclear. I will open issues for them.

**Comment by rovellipaolo** [comment]
About memoization, check my comment above.
The change was actually really nice, but was breaking the unit tests and I didn't want to block the rest of the pull request (i.e. multidex support, docker support, ...) for that.
Feel free to send the change again in a separate pull request but, of course, with the tests fixed. ;)

**Pull request merged**